packages/vscode-extension/src/debug/prerequisitesHandler.ts

```diff
@@ -109,82 +128,112 @@ export async function checkAndInstall(): Promise<Result<any, FxError>> {
+    await checkFailure(checkResults, progressBar);
+
     const checkPromises = [];
+    const progressMessages = [];
 
+    const parallelProgressHelper = new ParallelProgressHelper(progressBar);
     // backend extension
-    const backendExtensionPromise = resolveBackendExtension(depsManager, projectSettings);
-    if (backendExtensionPromise) {
-      checkPromises.push(backendExtensionPromise);
+    if (enabledCheckers.includes(Checker.AzureFunctionsExtension)) {
+      checkPromises.push(resolveBackendExtension(depsManager, parallelProgressHelper));
+      progressMessages.push(ProgressMessage.resolveBackendExtension);
     }
 
     // npm installs
-    if (ProjectSettingsHelper.isSpfx(projectSettings)) {
+    if (enabledCheckers.includes(Checker.SPFx)) {
       checkPromises.push(
         checkNpmInstall(
           Checker.SPFx,
           path.join(workspacePath, FolderName.SPFx),
-          "tab app (SPFx-based)"
+          NpmInstallDisplayName.SPFx,
+          parallelProgressHelper
         )
       );
-    } else {
-      if (ProjectSettingsHelper.includeFrontend(projectSettings)) {
-        checkPromises.push(
-          checkNpmInstall(
-            Checker.Frontend,
-            path.join(workspacePath, FolderName.Frontend),
-            "tab app (react-based)"
-          )
-        );
-      }
+      progressMessages.push(ProgressMessage.checkNpmInstall(NpmInstallDisplayName.SPFx));
+    }
 
-      if (ProjectSettingsHelper.includeBackend(projectSettings)) {
-        checkPromises.push(
-          checkNpmInstall(
-            Checker.Backend,
-            path.join(workspacePath, FolderName.Function),
-            "function app"
-          )
-        );
-      }
+    if (enabledCheckers.includes(Checker.Frontend)) {
+      checkPromises.push(
+        checkNpmInstall(
+          Checker.Frontend,
+          path.join(workspacePath, FolderName.Frontend),
+          NpmInstallDisplayName.Frontend,
+          parallelProgressHelper
+        )
+      );
+      progressMessages.push(ProgressMessage.checkNpmInstall(NpmInstallDisplayName.Frontend));
+    }
 
-      if (ProjectSettingsHelper.includeBot(projectSettings)) {
-        checkPromises.push(
-          checkNpmInstall(Checker.Bot, path.join(workspacePath, FolderName.Bot), "bot app")
-        );
-      }
+    if (enabledCheckers.includes(Checker.Backend)) {
+      checkPromises.push(
+        checkNpmInstall(
+          Checker.Backend,
+          path.join(workspacePath, FolderName.Function),
+          NpmInstallDisplayName.Function,
+          parallelProgressHelper
+        )
+      );
+      progressMessages.push(ProgressMessage.checkNpmInstall(NpmInstallDisplayName.Function));
     }
 
+    if (enabledCheckers.includes(Checker.Bot)) {
+      checkPromises.push(
+        checkNpmInstall(
+          Checker.Bot,
+          path.join(workspacePath, FolderName.Bot),
+          NpmInstallDisplayName.Bot,
+          parallelProgressHelper
```

Comment:
Is `parallelProgressHelper` only for `next()`?
Consider `checkNpmInstall().then(progress.next())`. Try to handle progress at top level as much as possible, instead of passing into each checker.